src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/VirtualDocumentFactoryBase.cs

```diff
@@ -0,0 +1,139 @@
+        /// CodeRemoteContentDefinition.CodeRemoteContentTypeName if needed.
+        /// </summary>
+        /// <param name="contentType">Content type registered with IContentTypeRegistryService</param>
+        /// <returns>Corresponding LSP/remote content type</returns>
+        protected virtual IContentType ConvertToLSPContentType(IContentType contentType) => contentType;
+
+        public override bool TryCreateFor(ITextBuffer hostDocumentBuffer, out VirtualDocument virtualDocument)
+        {
+            if (hostDocumentBuffer is null)
+            {
+                throw new ArgumentNullException(nameof(hostDocumentBuffer));
+            }
+
+            if (!hostDocumentBuffer.ContentType.IsOfType(HostDocumentContentTypeName))
+            {
+                // Another content type we don't care about.
+                virtualDocument = null;
+                return false;
+            }
+
+            var hostDocumentUri = _fileUriProvider.GetOrCreate(hostDocumentBuffer);
+
+            // E.g. Index.cshtml => Index.cshtml__virtual.html (for html), similar for other languages
+            var virtualLanguageFilePath = hostDocumentUri.GetAbsoluteOrUNCPath() + LanguageFileNameSuffix;
+            var virtualLanguageUri = new Uri(virtualLanguageFilePath);
+
+            var languageBuffer = _textBufferFactory.CreateTextBuffer();
+            _fileUriProvider.AddOrUpdate(languageBuffer, virtualLanguageUri);
+            languageBuffer.Properties.AddProperty(LSPConstants.ContainedLanguageMarker, true);
+
+            if (!(AdditionalLanguageBufferProperties is null))
+            {
+                foreach (KeyValuePair<object, object> keyValuePair in AdditionalLanguageBufferProperties)
+                {
+                    languageBuffer.Properties.AddProperty(keyValuePair.Key, keyValuePair.Value);
+                }
+            }
+
+            // Create a text document to trigger language server initialization for the contained language.
+            _textDocumentFactory.CreateTextDocument(languageBuffer, virtualLanguageFilePath);
+
+            languageBuffer.ChangeContentType(LanguageLSPContentType, editTag: null);
+
+            virtualDocument = CreateVirtualDocument(virtualLanguageUri, languageBuffer);
+            return true;
+        }
+
+        /// <summary>
+        /// Creates and returns specific virtual document instance
+        /// </summary>
+        /// <param name="uri">Virtual document URI</param>
+        /// <param name="textBuffer">Language text buffer</param>
+        /// <returns>Language-specific virtual document instance</returns>
+        protected abstract VirtualDocument CreateVirtualDocument(Uri uri, ITextBuffer textBuffer);
+
+        /// <summary>
+        /// Returns contained language content type name that's registered with IContentTypeRegisteryService
+        /// </summary>
+        protected abstract string LanguageContentTypeName { get; }
```

Comment:
Instead of `ConvertToLSPContentType` and `LanguageContentTypeName` it'd probably be better to just have a `LanguageContentType` property which does this under the covers. Even though most implementations will end up calling the content type registry service and even caching the content type it'll reduce the "special" casing implication of having the two

Reply:
Yeah, it was looking a bit messy to me as well. I'll go with your suggestion as code duplication will be minimal I guess. 

---
In reply to: [451977646](https://github.com/dotnet/aspnetcore-tooling/pull/2201#discussion_r451977646) [](ancestors = 451977646)